### PR TITLE
restructure logging and add silencing options per ZenPack

### PR DESCRIPTION
**Pull request by j053ph4**
- add basic logging to all classes
- log entries note the affected ZenPack
- added "verbose" and "level" parameters to "load_yaml"
- default logging is error, but can be overridden

### CODE REVIEW

File: ZenPacks/zenoss/ZenPackLib/lib/base/CatalogBase.py [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
-from ..utils import LOG
+from ..functions import LOG
+
 
 class CatalogBase(object):
     """Base class that implements cataloging a property"""
 
     # By Default there is no default catalog created.
     _catalogs = {}
+    LOG=LOG
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/base/CatalogBase.py]: is this being used anywhere?

File: ZenPacks/zenoss/ZenPackLib/lib/base/ComponentBase.py [hunk 2]
```diff
@@ -151,27 +150,27 @@ def setIdsInRelationship(self, relationship, ids):
...
-                LOG.error(
+                self.LOG.error(
                     "setIdsInRelationship (%s): No targets found matching "
                     "id=%s", relationship, id_)
 
                 continue
 
             if id_ in new_ids:
-                LOG.debug("Adding %s to %s" % (obj, relationship))
+                self.LOG.debug("Adding %s to %s" % (obj, relationship))
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/base/ComponentBase.py]: At some point we're going to have to be python 3 compatible, so use .format() instead of %.  

`"Adding {} to {}".format(obj, relationship)`

File: ZenPacks/zenoss/ZenPackLib/lib/base/ComponentBase.py [hunk 3]
```diff
@@ -151,27 +150,27 @@ def setIdsInRelationship(self, relationship, ids):
...
                 relationship.addRelation(obj)
 
                 # Index remote object. It might have a custom path reporter.
                 notify(IndexingEvent(obj, 'path', False))
             else:
-                LOG.debug("Removing %s from %s" % (obj, relationship))
+                self.LOG.debug("Removing %s from %s" % (obj, relationship))
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/base/ComponentBase.py]: At some point we're going to have to be python 3 compatible, so use .format() instead of %.

"Adding {} to {}".format(obj, relationship)

File: ZenPacks/zenoss/ZenPackLib/lib/base/ComponentBase.py [hunk 4]
```diff
@@ -282,8 +281,7 @@ def get_facets(self, root=None, streams=None, seen=None, depth=0, path=None, rec
                     for stream in streams:
                         recurse = any([pattern.match(relpath) for pattern in stream])
 
-                        LOG.log(9, "[%s] matching %s against %s: %s" % (root.meta_type, relpath, [x.pattern for x in stream], recurse))
-
+                        self.LOG.log(9, "[%s] matching %s against %s: %s" % (root.meta_type, relpath, [x.pattern for x in stream], recurse))
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/base/ComponentBase.py]: Same as above

File: ZenPacks/zenoss/ZenPackLib/lib/base/ZenPack.py [hunk 5]
```diff
@@ -27,13 +29,13 @@ def __init__(self, *args, **kwargs):
...
-                LOG.setLevel(logLevel)
+                self.LOG.propagate = False
+                self.LOG.setLevel(logLevel)
                 h = logging.StreamHandler(sys.stderr)
                 h.setFormatter(logging.Formatter('%(asctime)s %(levelname)s %(name)s: %(message)s'))
-                LOG.addHandler(h)
+                self.LOG.addHandler(h)
             else:
-                LOG.error("Unrecognized ZPL_LOG_LEVEL '%s'" %
+                self.LOG.error("Unrecognized ZPL_LOG_LEVEL '%s'" %
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/base/ZenPack.py]: Use .format() instead of %

File: ZenPacks/zenoss/ZenPackLib/lib/base/ZenPack.py [hunk 6]
```diff
@@ -52,21 +54,21 @@ def install(self, app):
                 try:
                     self.dmd.Devices.getOrganizer(dcspec.path)
                 except KeyError:
-                    LOG.info('Creating DeviceClass %s' % dcspec.path)
+                    self.LOG.info('Creating DeviceClass %s' % dcspec.path)
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/base/ZenPack.py]: .format() instead of %

File: ZenPacks/zenoss/ZenPackLib/lib/base/ZenPack.py [hunk 7]
```diff
@@ -52,21 +54,21 @@ def install(self, app):
...
                     app.dmd.Devices.createOrganizer(dcspec.path)
 
             dcObject = self.dmd.Devices.getOrganizer(dcspec.path)
             for zprop, value in dcspec.zProperties.iteritems():
                 if dcObject.getPropertyType(zprop) is None:
-                    LOG.error("Unable to set zProperty %s on %s (undefined zProperty)", zprop, dcspec.path)
+                    self.LOG.error("Unable to set zProperty %s on %s (undefined zProperty)", zprop, dcspec.path)
                     continue
-                LOG.info('Setting zProperty %s on %s' % (zprop, dcspec.path))
+                self.LOG.info('Setting zProperty %s on %s' % (zprop, dcspec.path))
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/base/ZenPack.py]: .format or send in to logger as params

File: ZenPacks/zenoss/ZenPackLib/lib/base/ZenPack.py [hunk 8]
```diff
@@ -52,21 +54,21 @@ def install(self, app):
...
                 dcObject.setZenProperty(zprop, value)
 
         # Load objects.xml now
         super(ZenPack, self).install(app)
         if self.NEW_COMPONENT_TYPES:
-            LOG.info('Adding %s relationships to existing devices' % self.id)
+            self.LOG.info('Adding %s relationships to existing devices' % self.id)
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/base/ZenPack.py]: .format()

File: ZenPacks/zenoss/ZenPackLib/lib/base/ZenPack.py [hunk 9]
```diff
@@ -144,17 +146,17 @@ def remove(self, app, leaveObjects=False):
             for catalog in self.GLOBAL_CATALOGS:
                 catObj = getattr(dc, catalog, None)
                 if catObj:
-                    LOG.info('Removing Catalog %s' % catalog)
+                    self.LOG.info('Removing Catalog %s' % catalog)
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/base/ZenPack.py]: .format()

File: ZenPacks/zenoss/ZenPackLib/lib/base/ZenPack.py [hunk 10]
```diff
@@ -144,17 +146,17 @@ def remove(self, app, leaveObjects=False):
...
                     dc._delObject(catalog)
 
             if self.NEW_COMPONENT_TYPES:
-                LOG.info('Removing %s components' % self.id)
+                self.LOG.info('Removing %s components' % self.id)
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/base/ZenPack.py]: .format()

File: ZenPacks/zenoss/ZenPackLib/lib/base/ZenPack.py [hunk 11]
```diff
@@ -166,7 +168,7 @@ def remove(self, app, leaveObjects=False):
                     Device._relations = tuple([x for x in Device._relations
                                                if x[0] not in self.NEW_RELATIONS[device_module_id]])
 
-                LOG.info('Removing %s relationships from existing devices.' % self.id)
+                self.LOG.info('Removing %s relationships from existing devices.' % self.id)
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/base/ZenPack.py]: .format() or pass as param

File: ZenPacks/zenoss/ZenPackLib/lib/base/ZenPack.py [hunk 12]
```diff
@@ -175,10 +177,10 @@ def remove(self, app, leaveObjects=False):
                     try:
                         app.dmd.Devices.getOrganizer(organizerPath)
                     except KeyError:
-                        LOG.warning('Unable to remove DeviceClass %s (not found)' % dcspec.path)
+                        self.LOG.warning('Unable to remove DeviceClass %s (not found)' % dcspec.path)
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/base/ZenPack.py]: .format() or pass as param

File: ZenPacks/zenoss/ZenPackLib/lib/base/ZenPack.py [hunk 13]
```diff
@@ -175,10 +177,10 @@ def remove(self, app, leaveObjects=False):
...
                         continue
 
-                    LOG.info('Removing DeviceClass %s' % dcspec.path)
+                    self.LOG.info('Removing DeviceClass %s' % dcspec.path)
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/base/ZenPack.py]: .format() or pass as param

File: ZenPacks/zenoss/ZenPackLib/lib/helpers/ZenPackLibLog.py [hunk 14]
```diff
@@ -0,0 +1,43 @@
+import logging
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/helpers/ZenPackLibLog.py]: Add copyright info

File: ZenPacks/zenoss/ZenPackLib/lib/helpers/ZenPackLibLog.py [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+class ZenPackLibLog(object):
+    ''''''
+    zenpacks = {}
+    defaultlog = new_log('zpl.zenpacklib')
+
+    def get_log(self, id):
+        if id in self.zenpacks.keys():
+            log = self.zenpacks.get(id).get('log')
+            self.defaultlog.info('get_log (zp): %s' % id)
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/helpers/ZenPackLibLog.py]: .format() or pass as param

File: ZenPacks/zenoss/ZenPackLib/lib/helpers/ZenPackLibLog.py [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+            return log
+        self.defaultlog.info('get_log (default): %s' % id)
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/helpers/ZenPackLibLog.py]: .format() or pass as param

File: ZenPacks/zenoss/ZenPackLib/lib/helpers/ZenPackLibLog.py [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+        return self.defaultlog
+
+    def add_log(self, id, quiet=True, level=0):
+        '''add a new log'''
+        self.defaultlog.info('%s verbose: %s level: %s' % (id, quiet, level))
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/helpers/ZenPackLibLog.py]: .format() or pass as param

File: ZenPacks/zenoss/ZenPackLib/lib/helpers/ZenPackLibLog.py [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+        id = str(id)
+        if id not in self.zenpacks.keys():
+            log = new_log(id)
+            log.setLevel(level)
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/helpers/ZenPackLibLog.py]: you could do this in one line

`log.setLevel('ERROR' if quiet else level)`

File: ZenPacks/zenoss/ZenPackLib/lib/helpers/ZenPackLibLog.py [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+            # set to error if quiet is True (default)
+            if quiet:
+                log.setLevel('ERROR')
+            self.zenpacks[id] = {'log': log, 'quiet': quiet, 'level':level}
+        self.defaultlog.info("ADDED %s (level %s) %s" % (id, log.level, log.handlers))
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/helpers/ZenPackLibLog.py]: .format() or pass as param

File: ZenPacks/zenoss/ZenPackLib/lib/spec/GraphDefinitionSpec.py [hunk 20]
```diff
@@ -21,7 +24,7 @@ def __init__(
             hasSummary=None,
             graphpoints=None,
             comments=None,
-            _source_location=None
+            _source_location=None,
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/spec/GraphDefinitionSpec.py]: In the other classes you added log as a parameter.  did you forget to put it in here?

File: ZenPacks/zenoss/ZenPackLib/lib/spec/GraphDefinitionSpec.py [hunk 21]
```diff
@@ -50,7 +53,7 @@ def __init__(
         :type comments: list(str)
         """
         super(GraphDefinitionSpec, self).__init__(_source_location=_source_location)
-
+        self.LOG=log
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/spec/GraphDefinitionSpec.py]: log is undefined

File: ZenPacks/zenoss/ZenPackLib/lib/spec/Spec.py [hunk 22]
```diff
@@ -190,8 +193,9 @@ def apply_data_defaults(self, dictionary, default_defaults=None, leave_defaults=
                         if i not in dictionary_params.keys():
                             dictionary_params[i] = j
 
-    def specs_from_param(self, spec_type, param_name, param_dict, apply_defaults=True, leave_defaults=False):
+    def specs_from_param(self, spec_type, param_name, param_dict, apply_defaults=True, leave_defaults=False, log=LOG):
         """Return a normalized dictionary of spec_type instances."""
+        #log.info('specs_from_param: %s', spec_type)
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/spec/Spec.py]: go ahead and delete

File: ZenPacks/zenoss/ZenPackLib/lib/spec/ZenPackSpec.py [hunk 23]
```diff
@@ -113,29 +115,31 @@ def __init__(
             zProperties=zProperties,
             classes=classes,
             class_relationships=class_relationships,
-            device_classes=device_classes)
-
+            device_classes=device_classes, log=self.LOG)
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/spec/ZenPackSpec.py]: for consistency, put log=self.LOG on its' own line

File: ZenPacks/zenoss/ZenPackLib/lib/spec/ZenPackSpec.py [hunk 24]
```diff
@@ -113,29 +115,31 @@ def __init__(
...
         self.name = name
+        self.LOG.debug("------ %s ------" % self.name)
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/spec/ZenPackSpec.py]: .format() or pass as param

File: ZenPacks/zenoss/ZenPackLib/lib/spec/ZenPackSpec.py [hunk 25]
```diff
@@ -113,29 +115,31 @@ def __init__(
...
+            ZPropertySpec, 'zProperties', zProperties, log=self.LOG)
 
         # Class Relationship Schema
-        self.class_relationships = []
+        self.class_relationships = self.specparams.class_relationships
         if class_relationships:
             if not isinstance(class_relationships, list):
                 raise ValueError("class_relationships must be a list, not a %s" % type(class_relationships))
-
+
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/spec/ZenPackSpec.py]: remove whitespace

File: ZenPacks/zenoss/ZenPackLib/lib/spec/ZenPackSpec.py [hunk 26]
```diff
@@ -217,7 +221,7 @@ def get_bases(cls, bases=[]):
             # Plumb _relations
             for relname, relationship in class_.relationships.iteritems():
                 if not relationship.schema:
-                    LOG.error("Removing invalid display config for relationship %s from  %s.%s" % (relname, self.name, class_.name))
+                    self.LOG.error("Removing invalid display config for relationship %s from  %s.%s" % (relname, self.name, class_.name))
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/spec/ZenPackSpec.py]: .format() or pass as param

File: ZenPacks/zenoss/ZenPackLib/lib/functions.py [hunk 27]
```diff
@@ -302,22 +318,38 @@ def relationships_from_yuml(yuml):
 
 
 def getZenossKeywords(klasses):
-    kwset = set()
+    kwset = set(['_properties', '_relations'])
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/functions.py]: Since this PR is for debug logging, can you pull this out into a separate PR since it's dealing with the keywords issue?

File: ZenPacks/zenoss/ZenPackLib/lib/tests/TestCase.py [hunk 28]
```diff
@@ -37,7 +39,7 @@ def afterSetUp(self):
         try:
             zenpack_module = importlib.import_module(self.zenpack_module_name)
         except Exception:
-            LOG.exception("Unable to load zenpack named '%s' - is it installed? (%s)", self.zenpack_module_name)
+            self.LOG.exception("Unable to load zenpack named '%s' - is it installed? (%s)", self.zenpack_module_name)
```
Comment (dbouchillon) [inline ZenPacks/zenoss/ZenPackLib/lib/tests/TestCase.py]: i think this needs another argument supplied to it or remove one of the formatters

File: ZenPacks/zenoss/ZenPackLib/lib/helpers/ZenPackLibLog.py (hunk 14 above)
Comment (j053ph4) [inline ZenPacks/zenoss/ZenPackLib/lib/helpers/ZenPackLibLog.py]: I need to do that for all the files, I was going to do it as a separate PR

File: ZenPacks/zenoss/ZenPackLib/lib/spec/GraphDefinitionSpec.py (hunk 20 above)
Comment (j053ph4) [inline ZenPacks/zenoss/ZenPackLib/lib/spec/GraphDefinitionSpec.py]: No I didn't b/c it kept failing when I did and worked when I didn't.  I just realized that the log=None parameter is already there a few lines up.  I'm setting the default like the others
